openquake/hazardlib/mfd/tapered_gr_mfd.py

```diff
@@ -0,0 +1,195 @@
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+# pylint: disable=missing-docstring,protected-access
+
+import math
+
+from openquake.hazardlib.mfd.base import BaseMFD
+from openquake.hazardlib.mfd.truncated_gr import TruncatedGRMFD
+
+
+class TaperedGRMFD(BaseMFD):
+    """
+    Tapered Gutenberg-Richter MFD is defined as a typical Truncated
+    Gutenberg-Richter MFD up to a ``corner_mag`` above which an exponential
+    taper is applied; see [ref] for details. This implementation is based on
+    the United States National Seismic Hazard Map Project code [url].
+
+    :param min_mag:
+        The lowest possible magnitude for this MFD. The first bin in the
+        :meth:`result histogram <get_annual_occurrence_rates>` will be aligned
+        to make its left border match this value.
+    :param max_mag:
+        The highest possible magnitude. The same as for ``min_mag``: the last
+        bin in the histogram will correspond to the magnitude value equal to
+        ``max_mag - bin_width / 2``.
+    :param bin_width:
+        A positive float value -- the width of a single histogram bin.
+
+    Values for ``min_mag`` and ``max_mag`` don't have to be aligned with
+    respect to ``bin_width``. They get rounded accordingly anyway so that
+    both are divisible by ``bin_width`` just before converting a function
+    to a histogram. See :meth:`_get_min_mag_and_num_bins`.
+    """
+
+    MODIFICATIONS = set(())
+
+    def __init__(self, min_mag: float, max_mag: float, corner_mag: float,
+                 bin_width: float, a_val: float, b_val: float):
+
+        self.min_mag = min_mag
+        self.max_mag = max_mag
+        self.corner_mag = corner_mag
+        self.bin_width = bin_width
+        self.a_val = a_val
+        self.b_val = b_val
+
+        self.beta = 2. / 3. * self.b_val
+
+        # constants from Mfds.java
+        self.SMALL_MO_MAG: float = 4.
+        self.TAPERED_LARGE_MAG: float = 9.05
+
+        self.min_mo = mag_to_mo(self.SMALL_MO_MAG)
+        self.large_mo = mag_to_mo(self.TAPERED_LARGE_MAG)
+        self.corner_mo = mag_to_mo(corner_mag)
+
+        # This class uses a Double Truncated GR distribution, and the
+        # rates are modified per the tapering.
+        self._dt_gr = TruncatedGRMFD(self.min_mag, self.max_mag,
+                                     self.bin_width, self.a_val, self.b_val)
+        self._get_min_mag_and_num_bins = self._dt_gr._get_min_mag_and_num_bins
+
+        self.check_constraints()
+
+    def check_constraints(self):
+        """
+        Checks the following constraints:
+
+        * Bin width is greater than 0.
+        * Minimum magnitude is positive.
+        * Maximum magnitude is greater than minimum magnitude
+          by at least one bin width (or equal to that value).
+        * Corner magnitude is greater than minimum magnitude
```

Comment:
Are you admitting that m corner can be equal to mmax?

Reply:
Yes.  My understanding is that the tapered GR mathematically has no upper bound, i.e. it is defined from a minimum magnitude on the left to infinity on the right (e.g., [Kagan and Schoenberg, 2001](https://escholarship.org/content/qt4d7477sq/qt4d7477sq.pdf). The corner magnitude is part of the characteristic shape of the MFD (not in the characteristic earthquake sense of the word).  

However Mmax as we use it is necessary for the discretization only, to put an upper limit on the size of earthquakes that we wish to consider in the analysis rather than those that are theoretically possible.  In the instances in which we use MFD in our own work (rather than the immediate motivation of implementing models made by others) this should give us some ability to specify a rather large Mmax as those events may be rare well above the corner magnitude.  Of course this can create additional problems with GMPEs outside of their calibration range, etc., but you can only fix one problem at a time...

Reply:
Thanks. I appreciate this but if you allow Mcorner == Mmax you might loose the part of the MFD covered by the truncation. Right?

Reply:
Yes, it should be equivalent to a truncated GR in this case.  Is this not the expected behaviour?

Reply:
That's fine with me (although I wouldn't certainly define a TaperedGR to get a TrunGR ... ahaha)